tests/unit/mathematics/linalg/SGLinalg_unittest.cc

```diff
@@ -0,0 +1,72 @@
+#include <shogun/lib/config.h>
+
+#include <shogun/mathematics/linalg/LinalgNamespace.h>
+
+#ifdef HAVE_VIENNACL
+#include <shogun/mathematics/linalg/LinalgBackendViennaCL.h>
+#endif
+
+#include <shogun/lib/SGVector.h>
+#include <gtest/gtest.h>
+
+using namespace shogun;
+
+TEST(LinalgRefactor, linalg_cpu_backend_dot)
+{
+	const index_t size = 10;
+	SGVector<int32_t> a(size), b(size);
+	a.set_const(1);
+	b.set_const(2);
+
+	auto result = linalgns::dot(a, b);
+
+	EXPECT_NEAR(result, 20, 1E-15);
+}
+
+TEST(LinalgRefactor, linalg_gpu_backend_dot_without_gpu_backend)
+{
+	const index_t size = 10;
+	SGVector<int32_t> a(size), b(size), a_gpu, b_gpu;
+	a.set_const(1);
+	b.set_const(2);
+
+	a_gpu.set(linalgns::to_gpu(a));
+	b_gpu.set(linalgns::to_gpu(b));
+
+	auto result = linalgns::dot(a, b);
+
+	EXPECT_NEAR(result, 20, 1E-15);
+}
+
+#ifdef HAVE_VIENNACL
+TEST(LinalgRefactor, linalg_gpu_backend_dot_with_gpu_backend)
+{
+	sg_linalg->set_gpu_backend(new LinalgBackendViennaCL());
+
+	const index_t size = 10;
+	SGVector<int32_t> a(size), b(size), a_gpu, b_gpu;
+	a.set_const(1);
+	b.set_const(2);
+
+	a_gpu.set(linalgns::to_gpu(a));
+	b_gpu.set(linalgns::to_gpu(b));
+
+	auto result = linalgns::dot(a, b);
+
+	EXPECT_NEAR(result, 20, 1E-15);
+
+}
+#endif
+
+TEST(LinalgRefactor, gpu_transfer_between_viennacl_and_cpu_backend)
+{
+	const index_t size = 10;
+	SGVector<int32_t> a(size), b, c;
+	a.range_fill(0);
```

Comment:
Please structure your unit tests a bit more:
 * group by linalg operations and distinguish between linalg::math_call and linalg::memory_call type of calls
 * Make sure you unit test things as the flags before and after gpu transfer
 * Make sure every line of the low level code is covered by a unit test